docker: Allow using a custom image name

leandro-lucarella-sociomantic opened the pull request:
Also, by default it will use `git config hub.upstream` or `TRAVIS_REPO_SLUG` to get a more meaningful default image name.

leandro-lucarella-sociomantic committed:
docker: Allow using a custom image name

Also, by default it will use `git config hub.upstream` or
`TRAVIS_REPO_SLUG` to get a more meaningful default image name.

leandro-lucarella-sociomantic commented:
This is mostly to be able to use beaver also locally to run builds through the defined docker image.

I wrote this simple `make` wrapper script for example:
```sh
#!/bin/sh
set -u

BEAVER_PATH=${BEAVER_PATH:-$(git config -f .gitmodules submodule.beaver.path)}

set -e

beaver="${BEAVER_PATH}/bin/beaver"
if test -x "$beaver"
then
	export DIST=$(lsb_release -cs) DMD="${DMD:-dmd1}" TRAVIS=1
	"$beaver" dlang install
	"$beaver" dlang make "$@"
else
	exec /usr/bin/make "$@"
fi
```

If all projects use `beaver` as the image name, is hard to use with more than one project.

mihails-strasuns-sociomantic commented on README.md:
```diff
@@ -56,7 +56,15 @@ The simplest use case would be:
+
+The image name will be taken from the `BEAVER_DOCKER_IMG` environment variable,
+if present. If not present it falls back to the result of `git config
+hub.upstream` (in case you are using the
```
Can't you just use `origin` remote to make it work even in absence of `git hub clone`?

nemanja-boric-sociomantic replied:
In that case I guess `TRAVIS_REPO_SLUG` would have to have a higher priority than `origin`, since both would exist in travis env.

leandro-lucarella-sociomantic replied:
Yeah, but it needs more work parsing URLs, something with which I already burned myself :D

leandro-lucarella-sociomantic replied:
But then again, you can always use a different way to get your `BEAVER_DOCKER_IMG` in your script, so I think there is very little gain in getting too smart, I only added `git config hub.upstream` because it was trivial to do so, but I would be more inclined to remove that other than adding more ways to autodiscover.

mihails-strasuns-sociomantic reviewed (commented):
Looks reasonable though I find single `workimg` image with all required dependencies to be more practical for local development :)

leandro-lucarella-sociomantic commented:
> Looks reasonable though I find single workimg image with all required dependencies to be more practical for local development :)

I like to have only the required dependencies installed because this way I can detect if an unwanted dependency was introduced more easily, and also because I could have different versions of different dependencies for different projects.

Also, this doesn't stop you from having a single `workimg` if you want it :)

leandro-lucarella-sociomantic committed:
Add an make wrapper for dlang projects

This wrapper will try to emulate a build in travis, by using beaver to
run insider a docker container and using the env var and matrixes
defined for travis.

leandro-lucarella-sociomantic commented:
I added a the make wrapper script, slightly improved to support even building the matrix as defined for travis. I think it might be generally useful to build a project without almost any previous setup.

mihails-strasuns-sociomantic commented:
> Also, this doesn't stop you from having a single workimg if you want it :)

Sure, it was more of a "looks good but I won't use it" comment :D

mihails-strasuns-sociomantic commented on tools/make-wrapper-dlang:
```diff
@@ -0,0 +1,56 @@
+#!/bin/sh
```
This script could use some nice pretty printing for progress but better to do it if it actually proves itself used.
